container/sagemaker/python_service.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+
+import importlib.util
+import re
+from collections import namedtuple
+
+import falcon
+import requests
+
+DEFAULT_CONTENT_TYPE = 'application/json'
+DEFAULT_ACCEPT_HEADER = 'application/json'
+CUSTOM_ATTRIBUTES = 'X-Amzn-SageMaker-Custom-Attributes'
+
+Context = namedtuple('Context',
+                     'model_name, model_version, method, rest_uri, grpc_port, '
+                     'custom_attributes, request_content_type, accept_header, content_length')
+
+
+class InvocationResource(object):
+
+    def __init__(self):
+        if 'TFS_GRPC_PORT' not in os.environ:
+            raise EnvironmentError('GRPC port not set')
+        if 'TFS_REST_PORT' not in os.environ:
+            raise EnvironmentError('REST API port not set.')
+
+        self._tfs_grpc_port = os.getenv('TFS_GRPC_PORT')
+        self._tfs_rest_port = os.getenv('TFS_REST_PORT')
+
+        self._handler, self._input_handler, self._output_handler = self._import_handlers()
+        self._handlers = self._make_handler(self._handler,
+                                            self._input_handler,
+                                            self._output_handler)
+
+    def on_post(self, req, res):
+        data, context = self._parse_request(req)
+        try:
+            prediction, content_type = self._handlers(data, context)
+            res.status = falcon.HTTP_200
+            res.body, res.content_type = prediction, content_type
+        except Exception as e:
+            raise Exception("Error in handling request: {}".format(str(e)))
+
+    def _import_handlers(self):
+        spec = importlib.util.spec_from_file_location('inference',
+                                                      '/opt/ml/model/script/inference.py')
+        inference = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(inference)
+
+        _custom_handler, _custom_input_handler, _custom_output_handler = None, None, None
+        try:
```

Comment:
this try/catch is ok. `if hasattr(inference, 'handler'):...` is better stylewise and perf wise.